nnfabrik/templates/transfer/transferred_trained_model.py

```diff
@@ -0,0 +1,310 @@
+import tempfile
+import torch
+import os
+import datajoint as dj
+import numpy as np
+from nnfabrik.templates.checkpoint import TrainedModelChkptBase
+from nnfabrik.utility.dj_helpers import gitlog, make_hash
+from nnfabrik.templates.trained_model import TrainedModelBase
+
+
+class TransferredTrainedModelBase(TrainedModelBase):
+    """
+    A modified version of TrainedModel table which enables step-wise table population given the
+    step specification. Refer to the corresponding example notebook for a demo.
+    """
+
+    table_comment = "Transferred trained models"
+
+    @property
+    def definition(self):
+        definition = """
+        # {table_comment}
+        transfer_step:                     int          # transfer step
+        -> self().model_table
+        -> self().dataset_table
+        -> self().trainer_table
+        -> self().seed_table
+        collapsed_history:                 varchar(64)
+        data_transfer:                     tinyint 
+        ---
+        comment='':                        varchar(768) # short description 
+        score:                             float        # loss
+        output:                            longblob     # trainer object's output
+        ->[nullable] self().user_table
+        trainedmodel_ts=CURRENT_TIMESTAMP: timestamp    # UTZ timestamp at time of insertion
+        """.format(
+            table_comment=self.table_comment
+        )
+        return definition
+
+    class ModelStorage(TrainedModelBase.ModelStorage):
+        pass
+
+    class DataStorage(dj.Part):
+        @property
+        def definition(self):
+            definition = """
+            # Contains the data generated by the transfer step, stored externally.
+            -> master
+            ---
+            transfer_data:            attach@{storage}
+            """.format(
+                storage=self._master.storage
+            )
+            return definition
+
+    class CollapsedHistory(dj.Part):
+        """
+        For the result of two or more transfer steps to be uniquely identifiable,
+        we compress its entire history (the keys of all previous steps) into a single hash (`collapsed_history`).
+        This table keeps track of this process and can be used to recursively retrieve the transfer history.
+        """
+
+        definition = """
+        collapsed_history:                 varchar(64)
+        model_fn:                     varchar(64)
+        model_hash:                   varchar(64)
+        dataset_fn:                   varchar(64)
+        dataset_hash:                 varchar(64)
+        trainer_fn:                   varchar(64)
+        trainer_hash:                 varchar(64)
+        seed:                         int
+        prev_collapsed_history:             varchar(64)
+        """
+
+        @classmethod
+        def add_entry(cls, key):
+            key = {
```

Comment:
consider doing
```python
key = dict(key) # make a copy
key['prev_collapsed_history'] = key.pop('collapsed_history')
```

Reply:
This won't work since `key` also contains no-primary keys.

Reply:
If you know a method to retrieve the primary keys of a table as a list, that would simplify the new version of this method very much.

Reply:
`Table.heading.primary_key` is what you are looking for.

Reply:
`Table._heading.primary_key` doesn't work, since the primary keys are not set here for some reason. 
I only renamed the parameter `key` to `entry` for now.

Reply:
Did you also try it without the underscore, i.e. `Table.heading.primary_key`?

Reply:
That did the trick. I'll update it now. Thanks!